# Section 01. 판다스(Pandas) 소개

## (1) 판다스의 데이터형 소개

- 판다스는 데이터 분석에 유용한 라이브러리중 하나입니다. 특히 웹 브라우저에 2차원 형태의 테이
블로 보여줄 수 있어 데이터 분석에서 많이 사용하는 라이브러리입니다. 판다스에서 제공하는 데
이터형에는 2가지가 있습니다. 첫째 series형, 둘째 DataFrame형이 있습니다. DataFrame은 테이
블 형태의 객체라고 생각하면 되고 Series는 테이블의 column(열)과 같은 개념으로 이해하면 됩니
다. 다음은 Series 객체와 DataFrame 객체의 특징입니다

## (2) 판다스 라이브러리 불러오기

- 판다스가 설치되어 있지 않다면, pip 를 이용해 판다스를 설치합니다. pip install + 패키지로 패키지
를 설치할 수 있습니다. 파일 탐색기 등에서 원하는 경로를 선택하고 주소 표시줄에 ‘cmd’라고 입
력한 후 명령 프롬프트가 나타나면 다음과 같이 입력합니다.

In [48]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [49]:
!pip install pandas

## (3) pandas 패키지 로드하기

- 패키지를 불러올때는 import 명령어를 사용합니다. import 명령 뒤에 ‘as pd’라고 입력하는 이유는
메소드를 사용할 때 ‘padas.Series..’형태로 입력한다면 다소 길게 입력해야 하기 때문에 pd라는 별
칭을 사용하여 pd.Series와 같은 형태로 짧게 입력할 수 있기 때문입니다.

In [50]:
import pandas as pd

# Section 02. pandas 의 Series 와 DataFrame

Series는 1차원의 1개의 column(열)이고 DataFrame은 2차원의 테이블 형태라고 생각하면 됩니
다. 이번 단락에서 Series와 DataFrame을 직접 생성하는 방법에 대해 알아보도록 하겠습니다.

## (1)데이터 프레임(DataFrame) 객체  

## Series 생성하기
- List 형태의 값을 Series 형태로 변환해보겠습니다. 리스트형으로 저장되어 있는 [1, 2, 3, 4, 5] 값을
Series메소드를 이용해 Series형으로 만듭니다

In [51]:
pd.Series([1, 2, 3, 4, 5])

0    1
1    2
2    3
3    4
4    5
dtype: int64

Series 형을 변수에 저장할 수 있습니다. 아래와 같이 변수명은 자유롭게 사용하면 됩니다.

In [52]:
pd_s = pd.Series([1, 2, 3, 4, 5])
pd_s

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [53]:
# type은 데이터의 타입을 확인하는 함수로 위의 데이터가 series형인지 아래와 같이 체크할 수 있습니다.

type(pd_s)

pandas.core.series.Series

## (2) 리스트를 판다스 Series형/DataFrame형으로 바꾸기

- 리스트 형태의 자료를 판다스 시리즈형과 데이터프레임 형태로 바꾸어 보겠습니다. 리스트 형태
로 갖고 있는 데이터를 다음과 같은 형태로 판다스의 시리즈와 데이터 프레임으로 변환할 수 있습
니다.

### <원본> 리스트  
출력 결과를 보면 알 수 있듯이 pd_l은 리스트형임을 알 수 있습니다.

In [54]:
pd_l=[1, 2, 3, 4, 5]
print(pd_l)
type(pd_l)

[1, 2, 3, 4, 5]


list

### <Series로 변환>  
pd_l을 Series형태로 변환하고 변수에 할당해 보도록 하겠습니다. pd_s1 변수에 Series 메소드를
이용하여 pd_l에 저장된 값을 판다스의 Series로 변환하고 저장합니다. print함수를 이용해 출력해
보면 인덱스와 Series형태로 잘 변환된 것을 알 수 있고 type을 출력해보면 판다스의 Series라고 표
시되는 것을 확인할 수 있습니다.

In [55]:
pd_s1 = pd.Series(pd_l)
print(pd_s1)
type(pd_s1)

0    1
1    2
2    3
3    4
4    5
dtype: int64


pandas.core.series.Series

### <DataFrame으로 변환>
원본 pd_l의 리스트를 데이터프레임으로 변환해 보겠습니다. 앞의 예제에서 Series 메소드를 사용
한 것과 같이 DataFrame 메소드를 사용하면 쉽게 데이터프레임으로 변환할 수 있습니다. pd_s2라
는 변수에 DataFrame메소드를 이용해 pd_l을 인자로 사용하여 데이터 프레임으로 변환합니다.
pd_s2를 실행해보면 데이터 프레임 형태로 표시되는 것을 확인할 수 있습니다. 실행 결과에서 가
장 윗줄에 표시된 0은 Column 이름을 의미합니다.

In [56]:
pd_s2 = pd.DataFrame(pd_l)
pd_s2

0
0  1
1  2
2  3
3  4
4  5

In [57]:
# type함수를 이용해 pd_s2의 데이터 타입을 확인해보면 다음과 같이 데이터프레임이라고 표시되는 것을 확인할 수 있습니다

type(pd_s2)

pandas.core.frame.DataFrame

### 데이터프레임에 열 추가하기
- pd_s2에 열 `Frame을 이용해 데이터를 2차원 형태로 표현할 수 있습니다.

In [58]:
pd_l=[1, 2, 3, 4, 5]
pd_s2 = pd.DataFrame(pd_l)
pd_s2['1열'] = pd_l
pd_s2

0  1열
0  1   1
1  2   2
2  3   3
3  4   4
4  5   5

In [59]:
pd_l=[1, 2, 3, 4, 5]
pd_l2=[11, 21, 31, 41, 51]

pd_s2 = pd.DataFrame(pd_l)
print(pd_s2)
pd_s2['1열'] = pd_l
pd_s2['2열'] = pd_l
pd_s2[3] = pd_l2
pd_s2[0] = pd_l2

pd_s2

   0
0  1
1  2
2  3
3  4
4  5


0  1열  2열   3
0  11   1   1  11
1  21   2   2  21
2  31   3   3  31
3  41   4   4  41
4  51   5   5  51

In [60]:
#DataFrame의 Column을 확인하고 싶다면, “.columns”을 이용해 확인합니다.

pd_s2.columns

Index([0, '1열', '2열', 3], dtype='object')

### 데이터프레임에 열 이름 변경하기
- 데이터프레임의 열 이름을 변경해보겠습니다. rename메소드를 이용해 열 이름을 변경할 수 있습
니다. 변수명.rename(columns={열이름:'변경할열이름'})형태로 입력합니다. 여기서는 “0”을 “0
열”로 변경합니다.

In [61]:
pd_s3 = pd_s2.rename(columns={0:'0열'}) # 0번 컬럼의 이름을 '0열'로 바꾸겠다
pd_s3

0열  1열  2열   3
0  11   1   1  11
1  21   2   2  21
2  31   3   3  31
3  41   4   4  41
4  51   5   5  51

In [62]:
pd_s3 = pd_s2.rename(columns={3:'3열'})

pd_s3.columns = ['first_column', 'second_column', '3', '4']
pd_s3

first_column  second_column  3   4
0            11              1  1  11
1            21              2  2  21
2            31              3  3  31
3            41              4  4  41
4            51              5  5  51

## (3) 2차원 리스트를 데이터프레임으로 만들기

### <실습1> 2차원 리스트형을 데이터프레임으로 만들기  
2차원 리스트를 데이터 프레임(DataFrame) 객체로 만들어 보겠습니다. 2차원 리스트로 values값
을 생성합니다. score라는 변수에 리스트 형태로 값을 할당하면 됩니다. print함수를 이용해 type을
출력해보면 list라고 표시됩니다

In [63]:
import pandas as pd

# 각각 행과 열이 됨
score = [['정약용',70,80,90],
['이순신',70,80,100],
['이율곡',75,95,90]]
print(type(score))
score

<class 'list'>


[['정약용', 70, 80, 90], ['이순신', 70, 80, 100], ['이율곡', 75, 95, 90]]

In [64]:
print(type(score))

<class 'list'>


### <실습2> 2차원 리스트를 DataFrame으로 만들기  
score_df라는 변수에 pd.DataFrame(score) 형태로 앞서 만든 score리스트를 데이터프레임으로
변환하여 할당합니다. DataFrame의 왼쪽 첫번째 열은 index입니다. index는 0번부터 시작됩니다.
DataFrame의 첫번째 행은 columns(열)이 되고 실제 Data가 저장된 부분은 values라고 합니다.

In [65]:
# 데이터 프레임 생성

score_df=pd.DataFrame(score)
print(type(score_df))
score_df

<class 'pandas.core.frame.DataFrame'>


0   1   2    3
0  정약용  70  80   90
1  이순신  70  80  100
2  이율곡  75  95   90

### <실습3> 열 이름 만들기  
“.columns”메소드를 이용하여 열 이름을 한꺼번에 만들거나 변경할 수 있습니다. 만약 특정한 열
이름만 변경할 때는 “rename” 메소드를 사용하면 됩니다.

In [66]:
score_df.columns = ['성명', '국어', '영어', '수학']
score_df

성명  국어  영어   수학
0  정약용  70  80   90
1  이순신  70  80  100
2  이율곡  75  95   90

### <실습4> 전체 열 이름 변경하기
열 이름을 다시 한 번 변경해보겠습니다. 앞서 배운 columns 메소드를 이용해 같은 방법으로 열 이 름을 변경하면 됩니다.

In [67]:
score_df.columns = ['성명1', '국어1', '영어1', '수학1']
score_df

성명1  국어1  영어1  수학1
0  정약용   70   80   90
1  이순신   70   80  100
2  이율곡   75   95   90

### <실습5> 특정 열 이름 변경
rename메소드를 이용해 특정 열 이름만 변경할 수 있습니다. ‘변수명.rename(columns = {'열이름 1':’변경할열이름1', '열이름2':변경할열이름2',......})’형태로 입력하면 됩니다. 아래 예제의 경우 ‘성명1’이라는 열 이름은 변경하지 않고 국어1, 영어1, 수학1을 각각 korean, english, math로 변경한다.

In [68]:
# 변경하고 싶은 일부 컬럼명만 변경할 수 있다

score_df_r=score_df.rename(columns = {'국어1':'korean', '영어1':'english', '수학1':'math'})
score_df_r

성명1  korean  english  math
0  정약용      70       80    90
1  이순신      70       80   100
2  이율곡      75       95    90

## (4)딕셔너리형을 DataFrame 으로 만들기

#### <실습1> 딕셔너리 만들기  
앞서 배운 예제에서 리스트를 이용해 DataFrame을 만들어 보았습니다. 이번에는 딕셔너리를 이용
해 DataFrame을 만드는 방법에 대해 설명하도록 하겠습니다. 앞의 예제에서 리스트를 먼저 만든
것과 마찬가지로 딕셔너리 데이터를 먼저 입력하고 score_dict 변수에 할당합니다.

In [69]:
score_dict = { '성명': ['정약용','이순신','이율곡'],
'국어': [70,80,90],
'영어': [70,80,100],
'수학': [75,95,90]}
print(score_dict,type(score_dict))

{'성명': ['정약용', '이순신', '이율곡'], '국어': [70, 80, 90], '영어': [70, 80, 100], '수학': [75, 95, 90]} <class 'dict'>


#### <실습2> 딕셔너리를 데이터프레임으로 만들기  
<실습1>에서 만든 딕셔너리를 이용해 데이터 프레임을 만들어보겠습니다. DataFrame메소드를
이용해 앞서 만든 예제에서 딕셔너리가 저장된 변수인 score_dict를 인자로 입력합니다. print함수
를 이용해 type을 출력해보면 DataFrame으로 설정된 것을 확인할 수 있고, score_dict_df를 실행해
보면 다음과 같이 데이터 프레임형태로 데이터가 표시됩니다.

    - 딕셔너리는 자동으로 컬럼이 지정됩니다!

In [70]:
score_dict_df = pd.DataFrame(score_dict)
print(type(score_dict_df))
score_dict_df

<class 'pandas.core.frame.DataFrame'>


성명  국어   영어  수학
0  정약용  70   70  75
1  이순신  80   80  95
2  이율곡  90  100  90

#### <실습3> 데이터 구조 확인하기  
Shape 속성을 이용해 데이터 구조를 확인할 수 있습니다. ‘변수명.shape’ 형태로 입력하면
score_dict_df는 3행 4열로 이루어져 있음을 확인할 수 있습니다.

In [71]:
score_dict_df.shape

(3, 4)

#### <실습4> 인덱스 확인하기  
index속성을 이용하면 데이터프레임의 행 인덱스의 시작과 끝, 증가값을 확인할 수 있습니다.
score_dict_df의 경우 인덱스는 0부터 시작해서 3으로 끝나고 1씩 증가하는 것을 알 수 있습니다.

In [72]:
score_dict_df.index # 행의 index

RangeIndex(start=0, stop=3, step=1)

#### <실습5> columns 속성으로 열 확인하기  
“.columns” 속성을 통해 데이터프레임이 어떤 열로 구성되어있는지 확인할 수 있습니다. 앞서 사용한 index속성과 마찬가지로 변수명.columns라고 입력하면 구성된 열과 데이터타입이 표시됩니다.

In [73]:
score_dict_df.columns # 컬럼 이름 목록 , object : 문자열

Index(['성명', '국어', '영어', '수학'], dtype='object')

#### <실습6> values 속성을 이용한 값 확인하기  
columns 속성을 이용해 열을 확인하는 것 뿐만 아니라 “.values”속성을 통해 값을 확인할 수도 있
습니다. ‘변수명.values’형태로 입력하고 확인해보면 다음과 같이 어떤 값으로 이루어져 있는지 표
시됩니다.

In [74]:
score_dict_df.values

array([['정약용', 70, 70, 75],
       ['이순신', 80, 80, 95],
       ['이율곡', 90, 100, 90]], dtype=object)

#### <실습7> 데이터형 변경하기  
데이터 분석 시 object형을 숫자형으로 변경해 보겠습니다.. 먼저 문자 형태의 숫자로 이루어진 열
을 가진 DataFrame을 정의하고 그 값을 n이라는 변수에 할당합니다. n.types속성을 이용해서 결과
를 표시해보면 col_str1과 col_str2 모두 object형인 것을 확인할 수 있습니다.

In [75]:
n=pd.DataFrame({'col_str1':['1','2','3','4'],
                'col_str2':['5','6','7','8']})
print(n)
n.dtypes

  col_str1 col_str2
0        1        5
1        2        6
2        3        7
3        4        8


col_str1    object
col_str2    object
dtype: object

#### <실습8> int, float형으로 변경하기  
object형인 데이터를 int나 float로 변경할 수 있는 방법은 여러가지가 있지만 여기서는 astype메소
드를 이용해 데이터 형을 변경해보도록하겠습니다. n_df라는 변수에 ‘변수명.astype({열이름:데이
터형,열이름:데이터형})’ 형태로 설정합니다. 변수명은 n이고 변경할 열 이름은 col_str1과 col_str2
이므로 각각 입력하고 콜론(:)으로 구분후 변경할 데이터 형인 int, float를 설정합니다. print문을 이
용해서 n_df를 출력하면 다음과 같이 결과가 표시됩니다. n_df.dtypes 속성으로 데이터형을 확인
해보면 int32, float64로 변경된 것을 알 수 있습니다.

In [76]:
n_df=n.astype({'col_str1':int,'col_str2':float})
print(n_df)
n_df.dtypes

   col_str1  col_str2
0         1       5.0
1         2       6.0
2         3       7.0
3         4       8.0


col_str1      int32
col_str2    float64
dtype: object

#### <여기서 잠깐> pandas의 데이터타입
dtypes 속성을 이용하면 각각의 column이 가지는 데이터타입을 확인할 수 있습니다. ‘데이터프레
임.dtypes’ 형태로 입력하면 됩니다. 앞서 사용한 ‘score_dict_df’의 데이터 타입을 확인해보면 다음
과 같이 표시됩니다.

In [77]:
score_dict_df.dtypes

성명    object
국어     int64
영어     int64
수학     int64
dtype: object

## (5) index를 특정 column으로 지정하기

- 인덱스는 보통 0으로 시작하는 숫자 형태로 구성됩니다. 인덱스는 꼭 숫자 0, 1, 2,.. 형태가 아니라 다른 형태로 변경할 수도 있습니다. 여기서는 인덱스를 다른 column으로 변경하는 방법에 대해 알아보도록 하겠습니다.

#### <실습1> 원본 DataFrame 확인하기  
앞서 사용한 예제 중에서 score_dict_df를 사용할 것입니다. score_dict_df를 실행해보면 다음과 같
이 결과가 표시됩니다. 인덱스가 0, 1, 2, 3,.. 형태로 표시되는 것을 확인할 수 있습니다.

In [78]:
score_dict_df

성명  국어   영어  수학
0  정약용  70   70  75
1  이순신  80   80  95
2  이율곡  90  100  90

#### <실습2> column을 index로 변경하기  
index값을 특정 column으로 변경해 보겠습니다. 여기서는 성명의 column 값을 사용하겠습니다

In [79]:
score_dict_df.index = score_dict_df['성명']
score_dict_df

성명  국어   영어  수학
성명                   
정약용  정약용  70   70  75
이순신  이순신  80   80  95
이율곡  이율곡  90  100  90

#### <실습3> 데이터 타입 확인하기  
type함수를 이용하여 score_dict_df['성명']의 타입을 확인해보면 Series라고 표시됩니다. Series는
컬럼(열)을 의미합니다. 컬럼을 지정할 때는 괄호( [ ] ) 안에 따옴표로 묶어 Column명을 표시입력
합니다.

In [80]:
score_dict_df.index = score_dict_df['성명']
type(score_dict_df['성명'])

pandas.core.series.Series

→ 데이터 프레임에서 하나만 가져왔기 때문에 데이터 타입이 series이다.

In [81]:
score_dict_df

성명  국어   영어  수학
성명                   
정약용  정약용  70   70  75
이순신  이순신  80   80  95
이율곡  이율곡  90  100  90

## (6) 행/열 삭제하기

drop함수를 사용하면 데이터프레임의 행이나 열을 삭제할 수 있습니다. 열이름이나 행 이름을 입
력하고 삭제할 대상이 행인지, 열인지 axis를 이용해 결정할 수 있습니다. 간단한 예제를 통해 데이
터프레임의 행이나 열을 삭제하는 방법에 대해 설명합니다.  

- axis=1 열 삭제
- axis=0 행 삭제

#### <실습1> 원본 확인하기  
앞서 사용한 데이터프레임을 그대로 사용할 것입니다. score_dict_df를 실행해보면 다음과 같은 결과가 표시되는 것을 알 수 있습니다. 이 데이터를 이용해 행과 열을 삭제하는 방법을 살펴보겠습니다.

In [82]:
score_dict_df

성명  국어   영어  수학
성명                   
정약용  정약용  70   70  75
이순신  이순신  80   80  95
이율곡  이율곡  90  100  90

#### <실습2> 열 삭제하기  
앞의 실습 예제에서 ‘성명’열을 인덱스로 변경하였기 때문에 성명 열이 두 개가 표시됩니다. 여기서
는 ‘성명’열을 삭제하도록 하겠습니다. drop함수를 사용하면 행이나 열을 삭제할 수 있습니다. “성
명” 열이 삭제된 것을 확인할 수 있습니다.

In [83]:
score_dict_df.drop(['성명'],axis=1)

국어   영어  수학
성명              
정약용  70   70  75
이순신  80   80  95
이율곡  90  100  90

#### <실습3> 행 삭제하기  
앞서 배운 예제에서 열을 삭제해보았으므로 이번에는 행을 삭제해보도록 하겠습니다. 표시되는
데이터 중에서 ‘이순신’에 해당하는 행을 삭제하겠습니다. 마찬가지로 drop함수를 이용해 삭제할
행 이름인 [‘이순신’]을 입력하고 행을 삭제할 것이기 때문에 axis는 0으로 설정합니다. 결과를 보면
다음과 같이 ‘이순신’ 행이 삭제되어 표시되는 것을 알수 있습니다.

In [84]:
score_dict_df.drop(['이순신'],axis=0)

성명  국어   영어  수학
성명                   
정약용  정약용  70   70  75
이율곡  이율곡  90  100  90

# Section 03. 판다스로 파일 다루기

앞의 단락에서 직접 Series및 DataFrame을 정의하는 방법과 데이터타입을 지정하고 행과 열을 추
가 또는 삭제하는 방법을 설명하였습니다. 이번 단락에서는 다른 프로그램에서 작성된 파일을 판
다스로 불러들여 사용하는 방법에 대해 알아보도록 하겠습니다. 주로 사용되는 파일은 ‘쉼표로 구
분된 텍스트(csv)’나 ‘텍스트파일(txt)’, ‘엑셀 파일(xlsx)’ 등이 있습니다.

### (1) pandas 로 csv/ excel 파일 읽고 쓰기

판다스에서 사용할 수 있는 파일의 종류와 파일을 읽어들일때, 저장할때 사용하는 함수에 대해 간단하게 정리해보도록 하겠습니다.

<파일 종류>
- csv(comma separated values) : 쉼표로 데이터를 구분해놓은 텍스트 파일
- csv 파일을 pandas로 처리할 경우.

<파일을 읽을 때 사용하는 함수>
- csv : pd.read_csv(‘파일명’)
- excel : pd.read_excel(‘파일명’)

    <파일을 저장 할 때 사용하는 함수>
- csv : pd.to_csv(‘파일명’)
- excel : pd.to_excel(‘파일명’)

### <참고>한글이 깨져서 표시된다면?
- csv 파일은 mac, linux, windows등 운영체제마다 문자 인코딩 설정이 다르기 때문에 나타나는 현
상입니다. 이때는 사용하는 운영체제 종류에 따라 cp949(완성형)이나 utf-8(유니코드)로 인코딩
방식을 지정해주어야 합니다

In [85]:
# windows 사용시 : utf-8

df = pd.read_csv('파일명', encoding='utf-8')


# macOS, linux 사용시 : cp949
#df = pd.read_csv('파일명', encoding='cp949')

FileNotFoundError: [Errno 2] No such file or directory: '파일명'

### (2) 데이터 소개: 타이타닉 예제

사용할 원본 데이터에 대해 간단하게 소개해 보도록 하겠습니다. Kaggle 내 Titanic 데이터입니다. 데이터 구조를 살펴보면 다음과 같은 정보를 확인할 수 있습니다.

- survival(생존 여부) : 0 = No, 1 = Yes
- pclass(좌석 등급) : 1 = 1st, 2 = 2nd, 3 = 3rd
- 사회적 계층을 대표함 : 1st = Upper, 2nd = Middle, 3rd = Lower
- sex(성별)
- age(나이) : 1 보다 작을 경우 분수로 나타냄, 추정치일 경우 xx.5 형식으로 나타냄.
- sibsp(형제 자매 / 배우자 수)
- parch(부모와 자식 수) : 어떤 아이들은 유모와 탑승했으며, 그 결과 0이 있음.
- ticket(티켓 번호)
- fare(요금)
- cabin(객실 번호)
- embarked(정박 항구) : C = Cherbourg, Q = Queenstown, S = Southampton

### <실습1> 웹페이지의 csv파일 가져오기


웹페이지에 있는 csv 파일을 데이터프레임으로 저장하자.
https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv
https://raw.githubusercontent.com/sh2orc/datascience/master/titanic.csv
혹은 주피터노트북 실행되고 있는 폴더나 구글 코랩을 사용시 드라이브 마운드한 폴더에
titanic.csv을 올리고 이를 불러와도 된다

In [86]:
import pandas as pd

#df = pd.read_csv( titanic.csv ) # 직접 폴더에 파일을 업로드하는 경우
df = pd.read_csv('https://raw.githubusercontent.com/sh2orc/datascience/master/titanic.csv')
df

PassengerId\tSurvived\tPclass\tName\tSex\tAge\tSibSp\tParch\tTicket\tFare\tCabin\tEmbarked
1\t0\t3\tBraund           Mr. Owen Harris\tmale\t22\t1\t0\tA/5 21171\t7...                                        
2\t1\t1\tCumings          Mrs. John Bradley (Florence Briggs Thayer)\tf...                                        
3\t1\t3\tHeikkinen        Miss. Laina\tfemale\t26\t0\t0\tSTON/O2. 31012...                                        
4\t1\t1\tFutrelle         Mrs. Jacques Heath (Lily May Peel)\tfemale\t3...                                        
5\t0\t3\tAllen            Mr. William Henry\tmale\t35\t0\t0\t373450\t8....                                        
...                                                                    ...                                        
152\t1\t1\tPears          Mrs. Thomas (Edith Wearne)\tfemale\t22\t1\t0\...                                        
153\t0\t3\tMeo            Mr. Alfonzo\tmale\t55.5\t0\t0\tA.5. 11206\t8....                                        
154\t0\t3\tvan Billiard   Mr. Austin Blyler\tmale\t40.5\t0\t2\tA/5. 851...                                        
155\t0\t3\tOlsen          Mr. Ole Martin\tmale\t\t0\t0\tFa 265302\t7.31...                                        
156\t0\t1\tWilliams       Mr. Charles Duane\tmale\t51\t0\t1\tPC 17597\t...                                        

[156 rows x 1 columns]

### <실습2> 탭을 기준으로 분리하기

- 앞서 불러들인 데이터는 가독성이 좋지 않아 탭을 기준을 보기 좋게 분리해 표시해보겠습니다.
csv파일을 불러올때 옵션 ‘sep='\t'’를 이용해보자. sep=’\t’라고 입력하면 탭을 기준으로 열이 구분
되어 표시됩니다. Passengerid, survived, pclass, name, sex, age 등으로 분리된 결과를 확인할 수
있습니다. 

In [87]:
url='https://raw.githubusercontent.com/sh2orc/datascience/master/titanic.csv'
titanic_url= pd.read_csv(url,sep='\t')
titanic_url

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
151          152         1       1   
152          153         0       3   
153          154         0       3   
154          155         0       3   
155          156         0       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
155                        Williams, Mr. Charles Duane    male  51.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
151      0            113776  66.6000    C2        S  
152      0        A.5. 11206   8.0500   NaN        S  
153      2          A/5. 851  14.5000   NaN        S  
154      0         Fa 265302   7.3125   NaN        S  
155      1          PC 17597  61.3792   NaN        C  

[156 rows x 12 columns]

### (3) 자주 사용되는 판다스의 함수 소개

- 데이터 분석에 필요한 판다스의 함수에 대해 알아보도록 하겠습니다. 자주 사용하는 함수들의 사
용 방법을 설명함께 예제로 다루어 보았습니다. 예제로 제시된 내용을 실습하면서 판다스 함수의
사용 방법에 대해 알아보겠습니다.

tail 함수는 head함수와는 반대로 맨 끝에서 5행의 데이터를 출력합니다. 개수를 지정하지 않은 경
우 끝의 5행의 데이터가 출력되지만 원하는 행수를 입력하면 입력한 행 수만큼의 데이터를 확인할
수 있습니다. 아래 예제는 tail 함수를 이용해 마지막 7건의 데이터를 출력한 것입니다.

In [88]:
titanic_url.tail(7)

PassengerId  Survived  Pclass                               Name     Sex  \
149          150         0       2  Byles, Rev. Thomas Roussel Davids    male   
150          151         0       2         Bateman, Rev. Robert James    male   
151          152         1       1  Pears, Mrs. Thomas (Edith Wearne)  female   
152          153         0       3                   Meo, Mr. Alfonzo    male   
153          154         0       3    van Billiard, Mr. Austin Blyler    male   
154          155         0       3              Olsen, Mr. Ole Martin    male   
155          156         0       1        Williams, Mr. Charles Duane    male   

      Age  SibSp  Parch       Ticket     Fare Cabin Embarked  
149  42.0      0      0       244310  13.0000   NaN        S  
150  51.0      0      0  S.O.P. 1166  12.5250   NaN        S  
151  22.0      1      0       113776  66.6000    C2        S  
152  55.5      0      0   A.5. 11206   8.0500   NaN        S  
153  40.5      0      2     A/5. 851  14.5000   NaN        S  
154   NaN      0      0    Fa 265302   7.3125   NaN        S  
155  51.0      0      1     PC 17597  61.3792   NaN        C

len 함수를 이용해서 데이터의 개수를 확인할 수 있습니다. len(데이터프레임명) 형태로 입력하면
총 몇 건의 데이터를 가지는지 확인할 수 있습니다. titanic_url의 경우 156건의 데이터가 있는 것을
확인할 수 있습니다.

In [89]:
len(titanic_url)

156

shape함수 : 데이터 프레임의 형태(shape) 알아보기 shape함수를 사용할 수 있습니다. 결과는 tuple 형태로 반
환되며, 첫번째는 row, 두번째는 column을 의미합니다.

In [90]:
titanic_url.shape

(156, 12)

info 는 기본적인 row(행)의 정보와 데이터 타입 등의 정보를 제공합니다. info 메소드는 주로 null
값과 데이터 타입을 볼 때 활용합니다.

In [91]:
titanic_url.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  156 non-null    int64  
 1   Survived     156 non-null    int64  
 2   Pclass       156 non-null    int64  
 3   Name         156 non-null    object 
 4   Sex          156 non-null    object 
 5   Age          126 non-null    float64
 6   SibSp        156 non-null    int64  
 7   Parch        156 non-null    int64  
 8   Ticket       156 non-null    object 
 9   Fare         156 non-null    float64
 10  Cabin        31 non-null     object 
 11  Embarked     155 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 14.8+ KB


→ RangeIndex : range함수를 썼다고 생각하면 된다. 0에서부터 리턴해준다는 의미  
→ pandas는 기본적으로 데이터를 object(문자열)형태로 불러온다

### (4) 데이터형 변경하기

데이터 분석을 위하여 데이터프레임의 값을 모두 float형으로 바꿔 보겠습니다. 혹은 선택적으로
다른 형식으로의 변경도 가능합니다. 데이터형을 변경하기 전에 데이터프레임이 어떤 데이터형을
가지는지 확인해보겠습니다.

### <실습1> 원본 데이터형식 확인하기
‘.dtypes’ 를 이용하여 한꺼번에 모든 컬럼의 데이터 타입을 확인해봅시다. titanic_url.dtypes 이라
고 입력하면 데이터프레임을 이루는 열인 Series의 데이터 형식을 확인할 수 있습니다.

In [92]:
titanic_url.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

### <실습2> to_numeric 속성을 이용한 데이터형 변경
가능한 모든 문자형을 숫자형으로 변환하고, 변환할 수 없는 것은 NaN으로 표시합니다.

In [93]:
titanic_url1 = titanic_url.apply(pd.to_numeric,errors = 'coerce')
titanic_url1

PassengerId  Survived  Pclass  Name  Sex   Age  SibSp  Parch    Ticket  \
0              1         0       3   NaN  NaN  22.0      1      0       NaN   
1              2         1       1   NaN  NaN  38.0      1      0       NaN   
2              3         1       3   NaN  NaN  26.0      0      0       NaN   
3              4         1       1   NaN  NaN  35.0      1      0  113803.0   
4              5         0       3   NaN  NaN  35.0      0      0  373450.0   
..           ...       ...     ...   ...  ...   ...    ...    ...       ...   
151          152         1       1   NaN  NaN  22.0      1      0  113776.0   
152          153         0       3   NaN  NaN  55.5      0      0       NaN   
153          154         0       3   NaN  NaN  40.5      0      2       NaN   
154          155         0       3   NaN  NaN   NaN      0      0       NaN   
155          156         0       1   NaN  NaN  51.0      0      1       NaN   

        Fare  Cabin  Embarked  
0     7.2500    NaN       NaN  
1    71.2833    NaN       NaN  
2     7.9250    NaN       NaN  
3    53.1000    NaN       NaN  
4     8.0500    NaN       NaN  
..       ...    ...       ...  
151  66.6000    NaN       NaN  
152   8.0500    NaN       NaN  
153  14.5000    NaN       NaN  
154   7.3125    NaN       NaN  
155  61.3792    NaN       NaN  

[156 rows x 12 columns]

### <실습3> 확인하기
dtypes 속성을 이용해 확인해보면 object형이었던 데이터가 모두 float 형태로 변경된 것을 확인할
수 있습니다. 문자열은 모두 NaN으로 처리됩니다.

In [94]:
titanic_url1.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name           float64
Sex            float64
Age            float64
SibSp            int64
Parch            int64
Ticket         float64
Fare           float64
Cabin          float64
Embarked       float64
dtype: object

### (5) 데이터프레임의 열(column)이름 변경하기

이번에는 데이터프레임을 구성하는 열 이름을 변경하는 방법에 대해 설명하도록 하겠습니다. 판
다스에서 제공하는 속성들과 메소드를 이용하면 열 이름을 쉽게 변경할 수 있습니다.

### <실습1> 열 이름 확인하기
columns 속성을 이용하여 데이터프레임이 갖고 있는 열 이름을 확인할 수 있습니다.
‘titanic_url.columns’라고 입력하면 다음과 같은 결과가 표시됩니다. 앞 단락에서 설명했지만
dtype이 object인것은 문자열을 의미합니다.

In [95]:
titanic_url.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### <실습2> Column(열) 이름 재정의하기
열이름을 한꺼번에 변경해 보겠습니다. 마찬가지로 columns 속성을 이용하여 변경할 수 있습니
다. 이 경우 정확하게 열의 개수가 일치해야 가능합니다.

In [96]:
new_columns = ['passengerId', 'survived', 'pclass', 'name', 'sex', 'age', 'sibsp',
'parch', 'ticket', 'fare', 'cabin', 'embarked']
titanic_url.columns = new_columns
titanic_url.columns

Index(['passengerId', 'survived', 'pclass', 'name', 'sex', 'age', 'sibsp',
       'parch', 'ticket', 'fare', 'cabin', 'embarked'],
      dtype='object')

In [97]:
titanic_url

passengerId  survived  pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
151          152         1       1   
152          153         0       3   
153          154         0       3   
154          155         0       3   
155          156         0       1   

                                                  name     sex   age  sibsp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
155                        Williams, Mr. Charles Duane    male  51.0      0   

     parch            ticket     fare cabin embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
151      0            113776  66.6000    C2        S  
152      0        A.5. 11206   8.0500   NaN        S  
153      2          A/5. 851  14.5000   NaN        S  
154      0         Fa 265302   7.3125   NaN        S  
155      1          PC 17597  61.3792   NaN        C  

[156 rows x 12 columns]

### (6) 데이터프레임의 행(index)의 이름을 만들고 변경해보자

열과 마찬가지로 index에 해당하는 행이름도 만들고 변경할 수 있습니다. 기본적으로 index를 변
경하지 않은 상태는 0부터 1씩 증가하는 숫자 형태지만 다른 열의 데이터를 이용하거나 직접 다른
값으로 변경도 가능합니다.

### <실습1> index(행) 출력하기
‘.index’ 속성을 사용하면 데이터프레임이 가지는 행의 시작 번호와 끝 번호 증가 값을 확인할 수 있
습니다. ‘titanic_url.index’라고 입력해보면 0부터 156까지 1씩 증가하는 형태인 것을 확인할 수 있
습니다.

- start=0, stop=156 : 0부터 155까지의 156개의 숫자
- step=1 : 1씩 증가

In [98]:
titanic_url.index

RangeIndex(start=0, stop=156, step=1)

### <실습2> values 속성 사용하기
인덱스를 제외하고 행만 출력할 때는 “.values”를 사용합니다. titanic_url.values라고 입력해보면
배열 형태로 indexr가 표시되어 나타납니다.

In [99]:
titanic_url.values

array([[1, 0, 3, ..., 7.25, nan, 'S'],
       [2, 1, 1, ..., 71.2833, 'C85', 'C'],
       [3, 1, 3, ..., 7.925, nan, 'S'],
       ...,
       [154, 0, 3, ..., 14.5, nan, 'S'],
       [155, 0, 3, ..., 7.3125, nan, 'S'],
       [156, 0, 1, ..., 61.3792, nan, 'C']], dtype=object)

### <실습3> 특정 열만 분리하기
- age 열만 분리하여 타입을 출력해보겠습니다. age라는 변수에 ‘titanic_url’ 데이터프레임의 열인
[‘age]를 입력하고 타입을 출력해보면 다음과 같은 결과가 표시됩니다

In [101]:
age = titanic_url['age'] # Series 객체: 1차원 배열(ndarray)과 호환
print(type(age))
age

<class 'pandas.core.series.Series'>


0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
151    22.0
152    55.5
153    40.5
154     NaN
155    51.0
Name: age, Length: 156, dtype: float64

### (7) 정렬하기
- 데이터 프레임을 다양한 형태로 정렬하는 방법에 대해 살펴보도록 하겠습니다. 인덱스를 기준으
로 정렬하거나 특정 column을 기준으로 정렬할 수 있습니다. 하나의 column 기준으로 정렬할수도
있지만 두 개 이상의 column기준으로 정렬하는 방법에 대해 알아보고 정렬할때 사용하는 메소드
들에 대한 내용도 알아보도록 하겠습니다.

### <실습1> sort_index를 이용한 인덱스 기준 오름차순 정렬하기
별다른 옵션을 설정하지 않은 경우 오름차순으로 정렬됩니다. 데이터 프레임의 인덱스를 기준으
로 오름차순 정렬해보도록 하겠습니다. 인덱스 기준 정렬 할때는 sort_index() 메소드를 사용하면
됩니다.

### .sort_index()를 이용한 정렬
인덱스에 따라 정렬
괄호안에 입력하지 않거나, ascending=True를 입력하면 오름차순 정렬
괄호안에 ascending=False를 입력하면 내림차순 정렬

In [103]:
titanic_url.sort_index()

passengerId  survived  pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
151          152         1       1   
152          153         0       3   
153          154         0       3   
154          155         0       3   
155          156         0       1   

                                                  name     sex   age  sibsp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
155                        Williams, Mr. Charles Duane    male  51.0      0   

     parch            ticket     fare cabin embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
151      0            113776  66.6000    C2        S  
152      0        A.5. 11206   8.0500   NaN        S  
153      2          A/5. 851  14.5000   NaN        S  
154      0         Fa 265302   7.3125   NaN        S  
155      1          PC 17597  61.3792   NaN        C  

[156 rows x 12 columns]

### <실습2> sort_index를 이용한 인덱스 기준 내림차순 정렬하기
인덱스를 기준으로 내림차순 정렬할때는 ‘acecending=False’라고 설정합니다. 다음과 같이
titanic_url.sort_index(ascending=False)라고 설정하고 결과를 확인하면 다음과 같이 표시됩니다.

In [105]:
titanic_url.sort_index(ascending=False)

passengerId  survived  pclass  \
155          156         0       1   
154          155         0       3   
153          154         0       3   
152          153         0       3   
151          152         1       1   
..           ...       ...     ...   
4              5         0       3   
3              4         1       1   
2              3         1       3   
1              2         1       1   
0              1         0       3   

                                                  name     sex   age  sibsp  \
155                        Williams, Mr. Charles Duane    male  51.0      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
..                                                 ...     ...   ...    ...   
4                             Allen, Mr. William Henry    male  35.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
0                              Braund, Mr. Owen Harris    male  22.0      1   

     parch            ticket     fare cabin embarked  
155      1          PC 17597  61.3792   NaN        C  
154      0         Fa 265302   7.3125   NaN        S  
153      2          A/5. 851  14.5000   NaN        S  
152      0        A.5. 11206   8.0500   NaN        S  
151      0            113776  66.6000    C2        S  
..     ...               ...      ...   ...      ...  
4        0            373450   8.0500   NaN        S  
3        0            113803  53.1000  C123        S  
2        0  STON/O2. 3101282   7.9250   NaN        S  
1        0          PC 17599  71.2833   C85        C  
0        0         A/5 21171   7.2500   NaN        S  

[156 rows x 12 columns]

### <실습3> 특정 column 기준으로 정렬하기
sort_values(by=’age’)라고 설정하면 age 컬럼에 따라 정렬된다. 괄호 안에 옵션을 입력하지 않거
나 ascending=True를 입력하면 오름차순 정렬된다. 괄호 안에 ascending=False를 입력하면 내림
차순 정렬된다.

In [106]:
titanic_url.sort_values(by='age') #오름차순 정렬
titanic_url.sort_values(by='age', ascending=False) #내림차순 정렬

passengerId  survived  pclass                            name     sex  \
96            97         0       1       Goldschmidt, Mr. George B    male   
116          117         0       3            Connors, Mr. Patrick    male   
33            34         0       2           Wheadon, Mr. Edward H    male   
54            55         0       1  Ostby, Mr. Engelhart Cornelius    male   
94            95         0       3               Coxon, Mr. Daniel    male   
..           ...       ...     ...                             ...     ...   
121          122         0       3      Moore, Mr. Leonard Charles    male   
126          127         0       3             McMahon, Mr. Martin    male   
128          129         1       3               Peter, Miss. Anna  female   
140          141         0       3   Boulos, Mrs. Joseph (Sultana)  female   
154          155         0       3           Olsen, Mr. Ole Martin    male   

      age  sibsp  parch      ticket     fare  cabin embarked  
96   71.0      0      0    PC 17754  34.6542     A5        C  
116  70.5      0      0      370369   7.7500    NaN        Q  
33   66.0      0      0  C.A. 24579  10.5000    NaN        S  
54   65.0      0      1      113509  61.9792    B30        C  
94   59.0      0      0      364500   7.2500    NaN        S  
..    ...    ...    ...         ...      ...    ...      ...  
121   NaN      0      0   A4. 54510   8.0500    NaN        S  
126   NaN      0      0      370372   7.7500    NaN        Q  
128   NaN      1      1        2668  22.3583  F E69        C  
140   NaN      0      2        2678  15.2458    NaN        C  
154   NaN      0      0   Fa 265302   7.3125    NaN        S  

[156 rows x 12 columns]

### <실습4> 여러 개의 열을 기준으로 정렬하기
.sort_values(by=[‘age’, ‘pclass’]) 형태로 정렬기준을 설정하면 ‘age’와 ‘pclass’ 열을 기준으로 정렬
된다. age가 첫 번째 정렬 기준이 되고 age가 같은 경우 pclass에 따라 정렬된다. 마찬가지로 괄호
안에 옵션을 설정하지 않거나 ascending=True를 입력하면 오름차순 정렬된다. 괄호 안에
ascending=False를 입력하면 내림차순으로 정렬된다.

In [107]:
titanic_url.sort_values(by=['age', 'pclass']) #오름차순 정렬
titanic_url.sort_values(by=['age', 'pclass'], ascending=False) #내림차순 정렬

passengerId  survived  pclass  \
96            97         0       1   
116          117         0       3   
33            34         0       2   
54            55         0       1   
94            95         0       3   
..           ...       ...     ...   
154          155         0       3   
17            18         1       2   
31            32         1       1   
55            56         1       1   
64            65         0       1   

                                               name     sex   age  sibsp  \
96                        Goldschmidt, Mr. George B    male  71.0      0   
116                            Connors, Mr. Patrick    male  70.5      0   
33                            Wheadon, Mr. Edward H    male  66.0      0   
54                   Ostby, Mr. Engelhart Cornelius    male  65.0      0   
94                                Coxon, Mr. Daniel    male  59.0      0   
..                                              ...     ...   ...    ...   
154                           Olsen, Mr. Ole Martin    male   NaN      0   
17                     Williams, Mr. Charles Eugene    male   NaN      0   
31   Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
55                                Woolner, Mr. Hugh    male   NaN      0   
64                            Stewart, Mr. Albert A    male   NaN      0   

     parch      ticket      fare cabin embarked  
96       0    PC 17754   34.6542    A5        C  
116      0      370369    7.7500   NaN        Q  
33       0  C.A. 24579   10.5000   NaN        S  
54       1      113509   61.9792   B30        C  
94       0      364500    7.2500   NaN        S  
..     ...         ...       ...   ...      ...  
154      0   Fa 265302    7.3125   NaN        S  
17       0      244373   13.0000   NaN        S  
31       0    PC 17569  146.5208   B78        C  
55       0       19947   35.5000   C52        S  
64       0    PC 17605   27.7208   NaN        C  

[156 rows x 12 columns]

## (8) 데이터 프레임 요소 접근 하기, 범위 선택하기, 읽기, 수정 loc iloc

데이터 프레임의 행을 원하는 만큼 가져와서 표시할 수 있습니다. 앞서 파이썬 기초에서 배운 문자
열의 슬라이싱 방법과도 비슷합니다. 이번 단락에서는 몇 가지 함수를 이용하여 원하는 만큼 행을
가져오는 방법에 대해 살펴보도록 하겠습니다

### <실습1> 행의 범위를 지정해서 (해당 행의 모든) 값을 가져오기
ex) titanic_url[a:b]
지정된 범위의 행(레코드)의 값들을 불러와 보겠습니다. 행을 슬라이싱 하는 방법으로 원하는 영역
까지 표시할 수 있습니다.

- titanic_url[x:y] : x번째부터 y번째까지 행을 추출
- titanic_url[::-1] : 행의 순서를 거꾸로 변경

titanic_url[0:5] # df.head() 동일한 결과
titanic_url[0:1]
titanic_url[::3] # 3개씩 건너띄어서 출력된다.
titanic_url[::-1] # 행의 순서를 거꾸로 변경

### <실습2> 단순 index에 대한 범위 선택
앞서 배운 내용을 이용해 단순 인덱스를 이용해 범위를 선택하면 다음과 같이 결과를 표시할 수 있
습니다. [:3]이라고 슬라이싱하여 표시할 수 있는데 첫 데이터부터 세번째 데이터까지 표시하면 0
부터 2까지 총 3개의 인덱스 값에 따른 데이터가 표시되어 나타납니다.

In [110]:
titanic_url[:3]

passengerId  survived  pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                name     sex   age  sibsp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   parch            ticket     fare cabin embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

### <여기서 잠깐> 아래 예시와 같은 방법으로 입력하면 KeyError가 발생합니다.
행의 인덱스를 이용하여 원하는 행을 추출할 경우 df[시작:끝:증가값] 형태로 작성해야 합니다. 시
작값과 종료값이 생략되더라도 콜론(:)반드시 입력해주어야 제대로 된 결과를 얻을 수 있습니다.

In [111]:
titanic_url[0]   # KeyError
titanic_url[3,1] # KeyError
titanic_url[3,:] # KeyError

KeyError: 0

### (9)인덱스 번호를 이용하여 값을 가져오자

### 인덱스 범위 선택  
- df.loc[행번호, 열] : 인덱스 이름으로 범위 선택, 명시적인 index.
- df.iloc[행번호, 열] : 인덱스 번호로 범위 선택, 암묵적인 index, 파이썬 스타일, int형 0 부터 시작
- df.ix[ ] : 최신 버전에서 삭제됨

### loc 속성 사용
- loc 속성을 이용하면 데이터 프레임의 인덱스를 이용해 원하는 행 만큼의 데이터를 사용하여 가져
올 수 있습니다. 하나의 행만 지정할 수도 있고, 여러 개의 행을 지정할 수도 있습니다. 기본형식은
다음과 같습니다.

### <기본형식>
titanic_url.loc[시작인덱스:종료인덱스:증가값]

### <실습1> 한 행만 표시하기
loc속성을 이용해서 0번째 행 인덱스의 값을 가져오면 다음과 같이 한 건의 데이터만 표시되는 것
을 확인할 수 있습니다. 0번째 인덱스 뿐만 아니라 데이터프레임에 존재하는 다른 인덱스 번호를
설정할수도 있습니다.

In [113]:
titanic_url.loc[0]

passengerId                          1
survived                             0
pclass                               3
name           Braund, Mr. Owen Harris
sex                               male
age                               22.0
sibsp                                1
parch                                0
ticket                       A/5 21171
fare                              7.25
cabin                              NaN
embarked                             S
Name: 0, dtype: object

### <실습2> 여러 개의 행 표시하기
아래 예제는 0번 인덱스부터 5번 인덱스까지 행 데이터를 가져와 표시한 것입니다. 슬라이싱 형태
로 데이터를 가져왔다면 0에서 4번 인덱스까지만 표시되겠지만 5번 인덱스까지 잘 표시되는 것을
확인할 수 있습니다.

In [114]:
titanic_url.loc[0:5]

passengerId  survived  pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                                name     sex   age  sibsp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   

   parch            ticket     fare cabin embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q

### <실습3> 증가값 설정하기

시작 인덱스와 종료인덱스를 설정하면 해당 범위의 인덱스를 가진 행이 표시됩니다. 뿐만 아니라
증가값을 설정하면 2개 단위 또는 3개 단위로 데이터를 건너뛰면서 표시할 수도 있습니다.

In [116]:
titanic_url.loc[0:10:2]

passengerId  survived  pclass  \
0             1         0       3   
2             3         1       3   
4             5         0       3   
6             7         0       1   
8             9         1       3   
10           11         1       3   

                                                 name     sex   age  sibsp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
4                            Allen, Mr. William Henry    male  35.0      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   

    parch            ticket     fare cabin embarked  
0       0         A/5 21171   7.2500   NaN        S  
2       0  STON/O2. 3101282   7.9250   NaN        S  
4       0            373450   8.0500   NaN        S  
6       0             17463  51.8625   E46        S  
8       2            347742  11.1333   NaN        S  
10      1           PP 9549  16.7000    G6        S

### <실습5> loc 속성의 인덱스, column 이름을 이용해 데이터 가져오기
- 자주 사용하는 기능으로 loc 속성과 인덱스번호, column 이름을 이용해 원하는 영역의 데이터를
가져와 표시할 수 있습니다. 0번 인덱스부터 마지막까지 ‘age’ column의 모든 행을 가져와서 표시
할 수 있습니다.

In [117]:
titanic_url.loc[:, 'age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
151    22.0
152    55.5
153    40.5
154     NaN
155    51.0
Name: age, Length: 156, dtype: float64

### <실습6> Age, Sex 컬럼의 모든 행을 선택해보자.
앞서 배운 방법과 비슷한 방법이지만 이번에는 두 개 이상의 column 데이터를 가져와서 표시하는
방법에 대해 알아보도록 하겠습니다. ‘titanic_url.loc[:, ['age', 'sex']]’라고 설정하면 첫번째 index부
터 마지막 index까지 age, sex column의 모든 데이터를 가져와서 표시합니다.

In [118]:
titanic_url.loc[:, ['age', 'sex']]

age     sex
0    22.0    male
1    38.0  female
2    26.0  female
3    35.0  female
4    35.0    male
..    ...     ...
151  22.0  female
152  55.5    male
153  40.5    male
154   NaN    male
155  51.0    male

[156 rows x 2 columns]

### <실습7> Age, sex 컬럼의 3부터 7까지 선택해보자.
이번에는 전체 인덱스가 아니라 특정 인덱스까지 필요한 column데이터를 가져와서 표시해보도록
하겠습니다. titanic_url.loc[3:7, ['Age', 'Sex']]라고 설정하면 3번 인덱스부터 7번 인덱스까지의 데이
터를 Age와 Sex column만 가져와서 표시할 수 있습니다.

In [119]:
titanic_url.loc[3:7, ['age', 'sex']]

age     sex
3  35.0  female
4  35.0    male
5   NaN    male
6  54.0    male
7   2.0    male

### <실습8> ‘Survived’ 컬럼에서 ‘Age’ 컬럼까지 2행부터 5행까지 선택해보자.
앞의 <실습7>에서 다룬 내용과 같은 내용이라고 생각하면 됩니다. titanic_url.loc[2:5,
'survived':'age']라고 설정하면 2번인덱스에서 5번 인덱스까지 ‘Survied’ column과 ‘age’ column의
데이터를 가져와서 표시합니다.

In [120]:
titanic_url.loc[2:5, 'survived':'age']

survived  pclass                                          name     sex  \
2         1       3                        Heikkinen, Miss. Laina  female   
3         1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4         0       3                      Allen, Mr. William Henry    male   
5         0       3                              Moran, Mr. James    male   

    age  
2  26.0  
3  35.0  
4  35.0  
5   NaN

### iloc 속성 사용

loc 속성과 비슷하지만 loc속성은 데이터프레임의 인덱스를 이용하는 것이라면 iloc속성은 데이터
프레임의 행 순서를 이용해 원하는 행을 가져올때 사용합니다. 현재 데이터프레임의 행 인덱스가
0부터 시작하고 순서도 0부터 시작하기 때문에 loc속성을 사용한 것과 비슷한 결과가 나타납니다.
기본 형식도 loc와 같다고 생각하면 됩니다.

### <기본형식>
titanic_url.iloc[시작:종료:증가]

### <실습1> iloc를 이용해 한 행만 표시하기
df.iloc[인덱스번호]형태로 입력하면 한 행의 데이터만 표시할 수 있습니다.

In [121]:
titanic_url.iloc[0]

passengerId                          1
survived                             0
pclass                               3
name           Braund, Mr. Owen Harris
sex                               male
age                               22.0
sibsp                                1
parch                                0
ticket                       A/5 21171
fare                              7.25
cabin                              NaN
embarked                             S
Name: 0, dtype: object

### <실습2> iloc를 이용해 0부터 5인덱스까지 표시하기
df.iloc[시작행번호:종료행번호-1] 형태로 입력하면 됩니다. loc 속성과 달리 iloc 속성은 [0:5]라고
설정한 경우 0행부터 4행까지의 데이터가 표시되어 나타납니다.

In [122]:
titanic_url.iloc[0:5]

passengerId  survived  pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                name     sex   age  sibsp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   parch            ticket     fare cabin embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### <실습3> iloc에 증가값 설정하기
df.iloc[시작행:종료행-1:증가값] 형태로 입력합니다. 아래의 실습 예제의 결과를 보면 0행부터 10
행-1까지 3씩 증가한 행이 표시되어 나타납니다.

In [123]:
titanic_url.iloc[0:10:3]

passengerId  survived  pclass  \
0            1         0       3   
3            4         1       1   
6            7         0       1   
9           10         1       2   

                                           name     sex   age  sibsp  parch  \
0                       Braund, Mr. Owen Harris    male  22.0      1      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
6                       McCarthy, Mr. Timothy J    male  54.0      0      0   
9           Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1      0   

      ticket     fare cabin embarked  
0  A/5 21171   7.2500   NaN        S  
3     113803  53.1000  C123        S  
6      17463  51.8625   E46        S  
9     237736  30.0708   NaN        C

### <실습4> 인덱스 번호로 범위 선택 iloc (position으로 색인)
첫 번째 컬럼에서 세 번째까지 모든 행을 선택해보자. ‘:’라고 설정했기 때문에 0번 인덱스부터 마지
막 인덱스까지 표시되어 나타납니다. [0,2]라고 설정한 것은 0번째 열과 2번째 열이 표시되어 나타
납니다.

In [124]:
titanic_url.iloc[:, [0, 2]]

passengerId  pclass
0              1       3
1              2       1
2              3       3
3              4       1
4              5       3
..           ...     ...
151          152       1
152          153       3
153          154       3
154          155       3
155          156       1

[156 rows x 2 columns]

### <실습5> 특정 행 범위, 열번호를 이용한 데이터 표시1

'1:5’라고 설정했기 때문에 두 번째 행부터 다섯 번째 행까지 표시되고 ‘[0,2]’라고 설정했으므로 첫
번째, 세 번째 열이 표시되어 나타납니다.

In [125]:
titanic_url.iloc[1:5, [0, 2]]

passengerId  pclass
1            2       1
2            3       3
3            4       1
4            5       3

### <실습6> 특정 행 범위, 열번호를 이용한 데이터 표시2
1:5라고 설정했으므로 두 번째 행부터 다섯 번째 행까지 표시되고 ‘1:4’라고 설정했기 때문에 두 번
째부터, 네 번째 열까지 선택되어 표시됩니다.

In [127]:
titanic_url.iloc[1:5, 1:4]

survived  pclass                                               name
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...
2         1       3                             Heikkinen, Miss. Laina
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)
4         0       3                           Allen, Mr. William Henry

### 컬럼을 기준으로 값을 가져오기

앞의 단락에서 loc 속성과 iloc속성을 이용하여 인덱스나 행 번호를 이용해 행 데이터를 가져오는
방법에 대해 설명하였습니다. 이번 단락에서는 column 번호와 column 이름 등을 이용해 특정
column을 가져와 표시하는 방법에 대해 설명합니다.

### <실습> len 함수를 사용하여 데이터 프레임의 열 개수 표시하기
len함수와 columns 속성을 사용해 다음 예시와 같이 설정하면 데이터 프레임의 column수가 표시
되어 나타납니다.

In [136]:
import pandas as pd
titanic_url=pd.read_csv('https://raw.githubusercontent.com/sh2orc/datascience/master/titanic.csv', titanic_url= pd.read_csv(url,sep='\t')
len(titanic_url.columns)

SyntaxError: invalid syntax (Temp/ipykernel_11772/2091721684.py, line 3)

In [133]:
titanic_url[titanic_url.columns[0]] # 0번 컬럼을 Series 객체로 추출, 1차원 배열

1\t0\t3\tBraund             Mr. Owen Harris\tmale\t22\t1\t0\tA/5 21171\t7...
2\t1\t1\tCumings            Mrs. John Bradley (Florence Briggs Thayer)\tf...
3\t1\t3\tHeikkinen          Miss. Laina\tfemale\t26\t0\t0\tSTON/O2. 31012...
4\t1\t1\tFutrelle           Mrs. Jacques Heath (Lily May Peel)\tfemale\t3...
5\t0\t3\tAllen              Mr. William Henry\tmale\t35\t0\t0\t373450\t8....
                                                 ...                        
152\t1\t1\tPears            Mrs. Thomas (Edith Wearne)\tfemale\t22\t1\t0\...
153\t0\t3\tMeo              Mr. Alfonzo\tmale\t55.5\t0\t0\tA.5. 11206\t8....
154\t0\t3\tvan Billiard     Mr. Austin Blyler\tmale\t40.5\t0\t2\tA/5. 851...
155\t0\t3\tOlsen            Mr. Ole Martin\tmale\t\t0\t0\tFa 265302\t7.31...
156\t0\t1\tWilliams         Mr. Charles Duane\tmale\t51\t0\t1\tPC 17597\t...
Name: PassengerId\tSurvived\tPclass\tName\tSex\tAge\tSibSp\tParch\tTicket\tFare\tCabin\tEmbarked, Length: 156, dtype: object

### <실습> column 번호 사용하기2
앞서 다룬 예제와 같은 내용입니다. 열번호를 변경하여 1열의 데이터를 가져와서 표시해보았습니
다. 마찬가지로 df[df.columns[열번호]] 형태로 입력하면 번호에 해당하는 열 이름과 인덱스가 함
께 표시되어 나타납니다.

In [134]:
titanic_url[titanic_url.columns[1]]

IndexError: index 1 is out of bounds for axis 0 with size 1

In [ ]:
7일차 6-2장 판다스와 데이터 분석 20페이지부터~